xarray/core/utils.py

```diff
@@ -577,19 +583,20 @@ class NDArrayMixin(NdimSizeLenMixin):
     """
 
     __slots__ = ()
+    array: Any  # TODO: what is this type, maybe ExplicitlyIndexed?
```

Comment:
Anything that implements `dtype` (returning `np.dtype`), `shape` (returning `tuple[int, ...]`) and `__getitem__` (not exactly sure what the return value needs to be, but it probbaly doesn't matter for this interface). I guess is some very basic version of a "duck numpy array," of the sort that is often used for loading data from different storage backends.

As an aside, the comment in the second paragraph of the docstring able overriding properties should probably be deleted: "A subclass should... set the `array` property and override one or more of `dtype`, `shape` and `__getitem__`."

Reply:
Ok, I think a Protocol would be the correct solution here.
Maybe a `SupportsMinimalArray`.
The return type of `__getitem__` could be Any for now, would be still an improvement.

Reply:
Maybe I'll do that in a different PR to keep the changes smaller. Quite sure such a change will lead to many many different problems that need fixing.